numpy/core/src/umath/matmul.c.src

```diff
@@ -0,0 +1,393 @@
+    npy_bool scalar_out = (dm == 1 && dp == 1);
+    npy_bool scalar_vec = (dn == 1 && (dp == 1 || dm == 1));
+    npy_bool too_big_for_blas = (dm > NPY_MAX_INT || dn > NPY_MAX_INT ||
+                                 dp >= NPY_MAX_INT);
+    npy_bool i1blasable = (is1_m == sz && BLAS_ABLE(dm, is1_n, sz)) ||
+                          (is1_n == sz && BLAS_ABLE(dn, is1_m, sz));
+    npy_bool i2blasable = (is2_n == sz && BLAS_ABLE(dn, is2_p, sz)) ||
+                          (is2_p == sz && BLAS_ABLE(dp, is2_n, sz));
+    npy_bool o_c_blasable = (os_p == sz) && BLAS_ABLE(dp, os_m, sz);
+    npy_bool o_f_blasable = (os_m == sz) && BLAS_ABLE(dm, os_p, sz);
+    npy_bool vector_matrix = ((dm == 1) && i2blasable);
+    npy_bool matrix_vector = ((dp == 1)  && i1blasable);
+#endif
+
+    if (dn == 0) {
+        /* No operand, need to zero the output */
+        for (iOuter = 0; iOuter < dOuter; iOuter++,
+                        args[0] += s0, args[1] += s1, args[2] += s2) {
+            npy_intp m, p;
+            char *op=args[2];
+            for (m = 0; m < dm; m++) {
+                for (p = 0; p < dp; p++) {
+#if @IS_COMPLEX@ == 1
+                    (*(@typ@ *)op).real = 0;
+                    (*(@typ@ *)op).imag = 0;
+#else
+                    *(@typ@ *)op = 0;
+#endif
+                    op  +=  os_p;
+                }
+                op  +=  os_m - (os_p * dp);
+            }
+        }
+        return;
+    }
+    for (iOuter = 0; iOuter < dOuter; iOuter++,
+                         args[0] += s0, args[1] += s1, args[2] += s2) {
+        char *ip1=args[0], *ip2=args[1], *op=args[2];
+#if @USEBLAS@ && defined(HAVE_CBLAS)
+        /*
+         * TODO: refactor this out to a inner_loop_selector, in
+         * PyUFunc_MatmulLoopSelector. But that call does not have access to
+         * n, m, p and strides.
+         */
+        if (too_big_for_blas) {
+            @TYPE@_matmul_inner_noblas(ip1, ip2, op, dm, dn, dp,
+                      is1_m, is1_n, is2_n, is2_p, os_m, os_p);
+        }
+        else if (special_case) {
+            /* Special case variants that have a 1 in the core dimensions */
+            if (scalar_out) {
+                /* row @ column, 1,1 output */
+                @TYPE@_dot(ip1, is1_n, ip2, is2_n, op, dn, NULL);
+            } else if (scalar_vec){
+                /*
+                 * 1,1d @ vector or vector @ 1,1d
+                 * could use cblas_Xaxy, but that requires 0ing output
+                 * and would not be faster (XXX prove it)
+                 */
+                @TYPE@_matmul_inner_noblas(ip1, ip2, op, dm, dn, dp,
+                          is1_m, is1_n, is2_n, is2_p, os_m, os_p);
+            } else if (vector_matrix) {
+                /* vector @ matrix, switch ip1, ip2, p and m */
+                @TYPE@_gemv((void *)ip2, is2_p, is2_n, (void *)ip1, is1_n,
+                            (void *)op, dp, dn);
+            } else if  (matrix_vector) {
```

Comment:
Perhaps rename to `blasable_matrix_vector`, and `vector_blasable_matrix` above?

Reply:
will do if there are more fixes

Reply:
done